### PR TITLE
Fix crash in gradient sampling

### LINKED ISSUE #3826: Second stop at 0% in gradient crashes compiler in web app
### Description

This typst code consistently crashes the compiler in the web app.
```typ
#rect(fill: gradient.linear((white, 0%), (white, 0%), (white, 100%)))
```

### Reproduction URL

_No response_

### Operating system

Web app

### Typst version

- [X] I am using the latest version of Typst

### PR DESCRIPTION
**Pull request by frozolotl**
Closes #3826.

CC: @Dherse

### CODE REVIEW

File: crates/typst/src/visualize/gradient.rs [hunk 1]
```diff
@@ -1249,7 +1249,11 @@ fn sample_stops(stops: &[(Color, Ratio)], mixing_space: ColorSpace, t: f64) -> C
 
     let (col_0, pos_0) = stops[low - 1];
     let (col_1, pos_1) = stops[low];
-    let t = (t - pos_0.get()) / (pos_1.get() - pos_0.get());
+    let mut delta = pos_1.get() - pos_0.get();
+    if delta == 0.0 {
```
Comment (laurmaedje) [inline crates/typst/src/visualize/gradient.rs:1253]: can we get into trouble if `delta` isn't equal to `0.0`, but very close to it?

Comment (Enivex) [reply crates/typst/src/visualize/gradient.rs:1253]: The current way this is dealt with seems somewhat arbitrary to me. Wouldn't it make more sense to simply make t either 0.0, 0.5 or 1.0 in the event that two stops are the same?

Maybe I misunderstand what's going on though.

Comment (laurmaedje) [reply crates/typst/src/visualize/gradient.rs:1253]: I agree. As-is it behaves as if the two stops were actually very far apart (the full `100% = 1.0`).

**Comment by Enivex** [comment]
https://discord.com/channels/1054443721975922748/1088371867913572452/1224684053870153848

Some minor discussion on discord

File: crates/typst/src/visualize/gradient.rs [hunk 2]
```diff
@@ -1249,7 +1249,11 @@ fn sample_stops(stops: &[(Color, Ratio)], mixing_space: ColorSpace, t: f64) -> C
...
+        delta = 1.0;
+    }
+    let t = (t - pos_0.get()) / delta;
```
Comment (laurmaedje) [inline crates/typst/src/visualize/gradient.rs:1256]: We can also deduplicate stops beforehand, but I think we should in any case clamp `t` to `0.0..=1.0` here. You never know what's gonna happen with floats.

**Comment by laurmaedje** [comment]
@frozolotl still on your radar?

**Comment by laurmaedje** [comment]
Closing due to inactivity. Feel free to reopen if you still want to work on this!

**Pull request closed without merging**